### PR TITLE
Weather app: toolbar and settings screen

### PR DESCRIPTION
**Pull request by jonholmberg**
Added a custom toolbar and a settings screen that lets you change the city so that it automatically updates the forecast when you return to the main screen.

Here are a few test id's if you desire (default is Copenhagen: 2618425):
Vejle: 2610613
London: 6058563
Helsinki: 658225
(An entire list can be found [here](http://openweathermap.org/current))

Toolbar features home/up button and menu with settings. It is hidden for the user when scrolling down and shown again when scrolling up.

Settings activity uses a custom delegate that stores the city in the shared preferences.

### CODE REVIEW

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/DetailActivity.kt [hunk 1]
```diff
@@ -21,11 +26,13 @@ class DetailActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_detail)
-        title = intent.getStringExtra(FORECAST_CITY_EXTRA_KEY)
+        initToolbar()
+        toolbarTitle = intent.getStringExtra(FORECAST_CITY_EXTRA_KEY)
+        enableHomeAsUp { onBackPressed() }
         bindForecast(intent.getSerializableExtra(FORECAST_EXTRA_KEY) as Forecast)
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/DetailActivity.kt]: What do you think will happen if the cast fails or we don't have a values stored in the bundle for `FORECAST_EXTRA_KEY`?

Comment (tudorgk) [reply jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/DetailActivity.kt]: _Hint: safe null cast_

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt [hunk 2]
```diff
@@ -3,32 +3,45 @@ package com.e_conomic.weatherapp.ui.activities
...
 
-    val COPENHAGEN_ID = "2618425"
+    override val toolbar by lazy {find<Toolbar>(R.id.toolbar)}
+
+    val cityId: Long by Preference(this, SettingsActivity.CITY_ID, SettingsActivity.DEFAULT_ID)
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        initToolbar()
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt]: How does a new developer know that he has to call `initToolbar` method before anything happens? I strongly encourage you to check out `abstract` classes 👍 .

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt [hunk 3]
```diff
@@ -3,32 +3,45 @@ package com.e_conomic.weatherapp.ui.activities
...
 
         forecastList.layoutManager = LinearLayoutManager(this)
+        attachToScroll(forecastList)
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt]: Same goes for this.

Consider writing an abstract function that expects a recyclerView. And get's called in the `onCreate()` method in the base abstract class

``` kotlin
abstract fun attachedRecyclerView(): RecyclerView
```

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt [hunk 4]
```diff
@@ -3,32 +3,45 @@ package com.e_conomic.weatherapp.ui.activities
...
-                            DetailActivity.FORECAST_CITY_EXTRA_KEY to forecastListResult.city)
+    override fun onResume() {
+        super.onResume()
+        loadForecast()
+    }
+
+    private fun loadForecast() =
+            doAsync {
+                val forecastListResult = RequestForecastCommand(cityId).execute()
+                uiThread {
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt:35]: We had some problems with this in the past in our sales app. Sometimes the activity might be `null` after the async task has been completed. Don't need to change anything yet, but keep in mind that this won't work in all situations. We'll dive into depth when you will be starting coding for our sales app.

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt [hunk 5]
```diff
@@ -3,32 +3,45 @@ package com.e_conomic.weatherapp.ui.activities
...
+                    forecastList.adapter = ForecastListAdapter(forecastListResult) { forecast ->
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/MainActivity.kt:36]: Nice touch with the listener block here.

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/utils/ToolbarManager.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import android.support.v7.widget.Toolbar
+import com.e_conomic.weatherapp.App
+import com.e_conomic.weatherapp.R
+import com.e_conomic.weatherapp.extensions.slideEnter
+import com.e_conomic.weatherapp.extensions.slideExit
+import com.e_conomic.weatherapp.ui.activities.SettingsActivity
+import org.jetbrains.anko.startActivity
+import org.jetbrains.anko.toast
+
+interface ToolbarManager {
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/utils/ToolbarManager.kt]: IMO this should be transformed into an `abstract class`. Even though you can have implemented methods in Kotlin interfaces, I think this is a wrong approach when considering that we need to initialise the manager in some way..

File: jlh/WeatherApp/app/src/main/res/layout/activity_detail.xml [hunk 7]
```diff
@@ -9,6 +9,8 @@
     android:paddingRight="@dimen/weather_item_horizontal_spacing"
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/res/layout/activity_detail.xml]: Why do we have these padding values here? 
screen shot 2017-04-03 at 11 10 46

**Review by tudorgk — CHANGES_REQUESTED** [review]
Looks good! 👍  I think some changes are need though
 so here's a 🌵  to get you started 😄 .

File: jlh/WeatherApp/app/src/main/res/layout/activity_detail.xml (hunk 7 above)
Comment (jonholmberg) [reply jlh/WeatherApp/app/src/main/res/layout/activity_detail.xml]: For a wonderful inconsistent layout :D No, I will change it haha :)

**Comment by jonholmberg** [comment]
@tudorgk I changed the `ToolbarManager` to an abstract class `ToolbarAppCompatActivity`. This now calls `initToolbar()` in the `onCreate()` method, but the content view is required so I have to call `setContentView()` before `super.onCreate()` in `MainActivity` and `DetailActivity` which is not so nice. Do you have a suggestion to make this better? I am very curious :) 

Also I didn't call `attachToScroll()` in `ToolbarAppCompatActivity.onCreate()` since it is not always used.

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/DetailActivity.kt [hunk 8]
```diff
@@ -1,35 +1,44 @@
...
     companion object {
         val FORECAST_EXTRA_KEY = "forecast"
         val FORECAST_CITY_EXTRA_KEY = "forecastCity"
     }
 
     override fun onCreate(savedInstanceState: Bundle?) {
-        super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_detail)
-        initToolbar()
+        super.onCreate(savedInstanceState)
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/activities/DetailActivity.kt]: Hmm maybe we can make this even prettier. Let's assume that we would always call `setContentView()` in `onCreate()`. So what we would actually need here is just the `int` value that represents the respective layout (in this case `R.layout.activity_detail`). If we create an abstract value name let's say `abstract val layoutId: Int` that will be requested as a parameter for `setContentView()`  in the abstract class' `onCreate` method, we can still call `super.onCreate` first. What do you think, does it get too complicated to follow?

File: jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/utils/ToolbarAppCompatActivity.kt [hunk 9]
```diff
@@ -11,15 +13,20 @@ import com.e_conomic.weatherapp.ui.activities.SettingsActivity
...
 
     var toolbarTitle: String
         get() = toolbar.title.toString()
         set(value) {
             toolbar.title = value
         }
 
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        initToolbar()
```
Comment (tudorgk) [inline jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/utils/ToolbarAppCompatActivity.kt]: You would call the `setContentView()` method here with the abstract value as a parameter.

**Review by tudorgk — APPROVED** [review]

**Pull request merged**